iterative_refinement/expectation_maximization.py

```diff
@@ -0,0 +1,435 @@
+    half_map_3d_final_1 : arr
+        Shape (n_pix, n_pix, n_pix)
+    half_map_3d_final_2 : arr
+        Shape (n_pix, n_pix, n_pix)
+    fsc_1d : arr
+        Final one dimensional fourier shell correlation.
+        Shape (n_pix // 2,)
+    """
+
+    def do_split(arr):
+        """
+        Split array into two halves along 0th axis.
+
+        Parameters
+        ----------
+        arr : arr
+            Shape (n_pix, n_pix, n_pix)
+
+        Returns
+        -------
+        arr1 : arr
+            Shape (n_pix // 2, n_pix, n_pix)
+        arr2: arr
+            Shape (n_pix // 2, n_pix, n_pix)
+        """
+        idx_half = arr.shape[0] // 2
+        arr_1, arr_2 = arr[:idx_half], arr[idx_half:]
+
+        if arr_1.shape[0] != arr_2.shape[0]:
+            arr_2 = arr[idx_half : 2 * idx_half]
+
+        return arr_1, arr_2
+
+    def do_build_ctf(ctf_params):
+        """
+        Build 2D array of evaluated CTFs from inputted
+        CTF parameters for each particle.
+
+        Parameters
+        ----------
+        ctf_params : list of dicts
+            Each dict contains CTF k,v pairs per particle.
+            Shape (n_particles,)
+
+        Returns
+        -------
+        ctfs : arr
+            Shape (n_ctfs, n_pix, n_pix)
+        """
+        n_ctfs = len(ctf_params)
+        ctfs = []
+
+        for i in range(n_ctfs):
+            ctfs.append(eval_ctf(**ctf_params[i]))
+
+        return ctfs
+
+    ctfs = do_build_ctf(ctf_info)
+    ctfs_1, ctfs_2 = do_split(ctfs)
+
+    # work in Fourier space. as particles stay in Fourier space the whole time.
+    # they are experimental measurements and are fixed in the algorithm
+    particles_1, particles_2 = do_split(particles)
+    particles_f_1 = do_fft(particles_1, d=3)
+    particles_f_2 = do_fft(particles_2, d=3)
+
+    n_pix = map_3d_init.shape[0]
+
+    max_n_iters = 7  # in practice programs using 3-10 iterations.
+
+    half_map_3d_r_1, half_map_3d_r_2 = map_3d_init, map_3d_init.copy()
+
+    for _ in range(max_n_iters):
+
+        # TODO: implement 3D fft using torch
+        half_map_3d_f_1 = do_fft(half_map_3d_r_1, d=3)
+        half_map_3d_f_2 = do_fft(half_map_3d_r_2, d=3)
+
+        def grid_SO3_uniform(n_rotations):
+            """
+            Generate a discrete set of uniformly distributed rotations
+            across SO(3).
+
+            Parameters
+            ----------
+            n_rotations : int
+                Number of rotations
+
+            Returns
+            -------
+            rots : arr
+                List of rotations.
+                Shape (n_rotations, 3, 3)
+            """
+            rots = np.ones((n_rotations, 3, 3))
+            return rots
+
+        n_rotations = 1000
+        rots = grid_SO3_uniform(n_rotations)
+
+        def generate_xy_plane(n_pix):
+            """
+            Generate xy plane.
+
+            Parameters
+            ----------
+            n_pix : int
+                Number of pixels
```

Comment:
`Number of pixels along one edge of the plane.`

It's confusing here since the grid is n_pix by n_pix and it gets squared. Technically there are n_pix**2 pixels. It's usually clear from context, but better to be explicit here.